### PR TITLE
Product Compare Backend / M2 Support

### LINKED ISSUE #1410: Product Compare / M2 Support
**Description**:
Compare Products generates a detailed, side-by-side comparison of two or more products.

**DESIGN**
mobile - zpl.io/VQjEX4P
desktop - zpl.io/VDEnEDr & zpl.io/VKRlR0q

**TODO**
- Create a compare page
- Add the compare icon in the header 
-- The link to the compare page should be added to the icon
- Style the page as per the design on mobile and desktop 
- Clear compare, price, and compare labels are translatable 
- Products cards and attributes are scrollable (check tdkz behavior on both mobile and desktop)
- Products can be removed from compare on click on X button
- Products can be added to cart on click on add to cart button 
- Product name and product image are linked with the respected products page
- On the clear compare button, remove all items from compare
- All customer groups can add items to compare 
- On mobile, attribute label is sticky and always visible  

**Expected result**

- Users can add products for comparison through PDP / PLP
- Comparison page created
- Any number of products can be added, but more than 4 introduces horizontal scroll bar as in TDKZ: https://www.technodom.kz/compare
- 2 products are visible on mobile, the third is partially visible to indicate scrolling ability
- Setting in Admin → Stores → Attributes → Product → {attribute} → Storefront properties → Comparable on Storefront is respected
- Able to compare configurable products, default variant is selected
- Header extended with icon leading to compare page. 
- Responsive

### PR DESCRIPTION
**Pull request by zans-laksa**
Fixes scandipwa/scandipwa#1410

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    },
+    "authors": [
+        {
+            "name": "Alfreds Genkins",
+            "email": "alfreds@scandiweb.com"
+        },
+        {
+            "name": "Lauris Grosu",
+            "email": "lauris.grosu@scandiweb.com"
+        }
```
Comment (alfredsgenkins) [inline composer.json]: We are not authors anymore, you can safely remove us and put your name there :D

File: src/Model/Resolver/ClearCompareProducts.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            $collection->setVisitorId($quoteIdMask);
+        }
+
+        $store = $this->storeManager->getStore();
+        $collection->setStore($store);
+
+        try {
+            // This loads items but throws undefined method exception which can be ignored
+            $collection->load();
+        } catch (\Exception $e) {}
```
Comment (alfredsgenkins) [inline src/Model/Resolver/ClearCompareProducts.php]: Interesting! It does not clear, if you don't load it?

File: src/Model/Resolver/CompareProductsResolver.php [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        } catch (\Exception $e) {}
+
+        $count = $collection->count();
+        $products = [];
+
+        if ($count) {
+            $productIds = $collection->getProductIds();
+
+            foreach ($productIds as $productId) {
+                $item = $this->productDataProvider->getProductDataById((int)$productId);
```
Comment (alfredsgenkins) [inline src/Model/Resolver/CompareProductsResolver.php]: This function does not seem legit. Becuase it most probably loads stuff one-by-one per product. This is inefficent.

File: src/Model/Resolver/CompareProductsResolver.php [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        try {
+            // This loads items but throws undefined method exception which can be ignored
+            $collection->load();
+        } catch (\Exception $e) {}
+
+        $count = $collection->count();
+        $products = [];
+
+        if ($count) {
+            $productIds = $collection->getProductIds();
```
Comment (alfredsgenkins) [inline src/Model/Resolver/CompareProductsResolver.php]: I would use these Product IDs with something similar to this: https://github.com/scandipwa/quote-graphql/blob/master/src/Model/Resolver/ProductResolver.php#L92:L114

The `$this->postProcessor->process` has a fourth argument https://github.com/scandipwa/performance/blob/master/src/Model/Resolver/Products/DataPostProcessor.php#L47 I would pass the is_compare there and then I would add a rewrite for this method https://github.com/scandipwa/performance/blob/master/src/Model/Resolver/Products/DataPostProcessor/Attributes.php#L290:L302 (in performance package or here) to load different attributes for compare products.

File: src/Model/Resolver/ClearCompareProducts.php [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        if ($customerId) {
+            $collection->setCustomerId($customerId);
+        } elseif ($guestCardId) {
+            $quoteIdMask = $this->quoteIdMaskFactory
+                ->create()
+                ->load($guestCardId, 'masked_id')
+                ->getQuoteId();
+
+            $collection->setVisitorId($quoteIdMask);
+        }
```
Comment (alfredsgenkins) [inline src/Model/Resolver/ClearCompareProducts.php]: This logic seems common to most files, consider moving it into helper!

File: src/etc/schema.graphqls [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    comparableAttributes: [AttributeWithValue] @resolver(class: "\\ScandiPWA\\CompareGraphQl\\Model\\Resolver\\ComparableAttributesResolver")
+}
+
+type AttributeWithValue {
+    attribute_id: Int
+    attribute_label: String
+    attribute_code: String
+    attribute_value: String
+    attribute_type: String
+}
```
Comment (alfredsgenkins) [inline src/etc/schema.graphqls]: I think implementing comments above we can survive with old arguments.

**Review by alfredsgenkins — CHANGES_REQUESTED** [review]

File: composer.json (hunk 1 above)
Comment (zans-laksa) [reply composer.json]: Done.

File: src/Model/Resolver/ClearCompareProducts.php (hunk 2 above)
Comment (zans-laksa) [reply src/Model/Resolver/ClearCompareProducts.php]: It does. Fixed it.

File: src/Model/Resolver/CompareProductsResolver.php (hunk 3 above)
Comment (zans-laksa) [reply src/Model/Resolver/CompareProductsResolver.php]: Fixed.

File: src/Model/Resolver/CompareProductsResolver.php (hunk 4 above)
Comment (zans-laksa) [reply src/Model/Resolver/CompareProductsResolver.php]: Fixed.

File: src/Model/Resolver/ClearCompareProducts.php (hunk 5 above)
Comment (zans-laksa) [reply src/Model/Resolver/ClearCompareProducts.php]: Fixed.

File: src/etc/schema.graphqls (hunk 6 above)
Comment (zans-laksa) [reply src/etc/schema.graphqls]: Fixed.

**Pull request merged**